pkg/controller/instance/instance_controller.go

```diff
@@ -348,3 +381,51 @@ func parameterDifference(old, new map[string]string) map[string]string {
 
 	return diff
 }
+
+func contains(values []string, s string) bool {
+	for _, value := range values {
+		if value == s {
+			return true
+		}
+	}
+	return false
+}
+
+func remove(values []string, s string) (result []string) {
+	for _, value := range values {
+		if value == s {
+			continue
+		}
+		result = append(result, value)
+	}
+	return
+}
+
+// maybeAddFinalizer adds the cleanup finalizer to an instance if the finalizer
+// hasn't been added yet, the instance has a cleanup plan and the cleanup plan
+// didn't run yet.
+func maybeAddFinalizer(instance *kudov1beta1.Instance) {
+	if !contains(instance.ObjectMeta.Finalizers, instanceCleanupFinalizerName) {
+		if planStatus := instance.PlanStatus(v1beta1.CleanupPlanName); planStatus != nil {
+			// avoid adding a finalizer again if a reconciliation is requested
+			// after it has just been removed but the instance isn't deleted yet
+			if planStatus.Status == v1beta1.ExecutionNeverRun {
+				log.Printf("InstanceController: Adding finalizer on instance %s/%s", instance.Namespace, instance.Name)
+				instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, instanceCleanupFinalizerName)
+			}
+		}
+	}
+}
+
+// maybeRemoveFinalizer removes the cleanup finalizer of an instance if it has
+// been added, the instance has a cleanup plan and the cleanup plan completed.
+func maybeRemoveFinalizer(instance *kudov1beta1.Instance) {
+	if contains(instance.ObjectMeta.Finalizers, instanceCleanupFinalizerName) {
+		if planStatus := instance.PlanStatus(v1beta1.CleanupPlanName); planStatus != nil {
```

Comment:
Maybe we should explicitly check if the operatorVersion has a cleanupPlan, and if not, just remove the finalizer. This is an edge case if an operatorVersion removes a `cleanup` plan that has previously existed.
If we don't do this, the finalizer is never removed and the object never deleted.

Reply:
Then we would remove the finalizer while a cleanup plan is running. Removing the finalizer will result in a deletion of the instance. We have to ensure that the finalizer is only removed _after_ the cleanup plan finished.